### PR TITLE
t/92: Introduce UI Controller class

### LINKED ISSUE #92: Introduce UI Controller class
Controllers, along with Views, create a tree–like structure of the UI. They hold the logic responsible for binding features to models and provide an API for dynamic view allocation.

Requires #93.

### PR DESCRIPTION
**Pull request by oleq**
Fixes #92.

### CODE REVIEW

File: tests/collection/collection.js [hunk 1]
```diff
@@ -10,8 +10,10 @@ const modules = bender.amd.require( 'collection', 'ckeditorerror' );
 bender.tools.createSinonSandbox();
 
 function getItem( id, idProperty ) {
+	idProperty = idProperty || 'id';
```
Comment (Reinmar) [inline tests/collection/collection.js]: :goat:

File: src/ui/controller.js [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+		}
+
+		/**
+		 * Returns a child controller from one of the {@link #collections} at given `index`.
+		 *
+		 * @param {String} collectionName One of {@link #collections} the child should be retrieved from.
+		 * @param {Number} [index] An index of desired controller.
+		 * @returns {Controller} A child controller instance.
+		 */
+		getChild( collectionName, index ) {
```
Comment (Reinmar) [inline src/ui/controller.js]: Seems to be unnecessary now as you can use `controller.collections.get( collectionName ).get( index )`.

File: src/ui/controller.js [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+			 */
+			this.view = view || null;
+
+			/**
+			 * A collection of {@link ControllerCollection} instances containing
+			 * child controllers.
+			 *
+			 * @property {Collection}
+			 */
+			this.collections = new Collection( {
```
Comment (Reinmar) [inline src/ui/controller.js]: Can't we call them `children`? I know that items of this collection are not children yet, but it won't look that bad:

```
collection.children.get( 'buttons' ).get( 0 );
```

File: src/ui/controllercollection.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ * Manages UI Controllers.
+ *
+ * @class ControllerCollection
+ * @extends Collection
+ */
+CKEDITOR.define( [
+	'collection',
+	'ckeditorerror'
+], ( Collection, CKEditorError ) => {
+	class ControllerCollection extends Collection {
```
Comment (Reinmar) [inline src/ui/controllercollection.js]: Is this class really necessary? Maybe we can add `opts.name` to the `Collection` class. Would make less code.

The counterargument is that _perhaps_ this class will get more behaviour in the future. But as we don't have any now, maybe let's KISS.

Comment (Reinmar) [inline src/ui/controllercollection.js]: Actually, I already have an idea for a customised behaviour that this class should have. I think that `Controller.addChild()` and `Controller.removeChild()` should be moved to `ControllerCollection.add()` and `ControllerCollection.remove()`. Then we would have this:

```
this.children.get( 'buttons' ).add( someButton )
    .then( ... ); // yep, add() would remove a promise as `Controller.addChild() does now
```

Comment (Reinmar) [inline src/ui/controllercollection.js]: Of course, that would mean that creating the buttons collection you would need to pass the controller instance:

```
this.children.add( new ControllerCollection( this, 'buttons' ) );
```

Comment (Reinmar) [inline src/ui/controllercollection.js]: Or without passing `this` to `ControllerCollection` – the controller could listen on the `ControllerCollection#add/remove` events.

File: tests/collection/collection.js (hunk 1 above)
Comment (oleq) [inline tests/collection/collection.js]: AFAIR it failed in JSCS.

**Comment by oleq** [comment]
I think the following is very confusing

```
this.children.get( 'buttons' ).get( 0 );
```

so I'm for `this.collections` rather than `this.children`. But if you insist, it's easy to change.

Anyway, I managed to implement the following syntax

```
this.collections.add( new ControllerCollection( 'buttons' ) );
this.collections.get( 'buttons' ).add( someButton )
    .then( ... );
```

so `ControllerCollection#add` returns promise and initialises the child controller when parent (`this`) is `ready == true`. It's far less code than `Controller#(add,remove,get)Child` and does exactly the same. 

The magic behind `new ControllerCollection( 'buttons' )` without passing `this` is in the `this.collections.on( 'add', ... )` listener, which sets `ControllerCollection#parent` property to `this`.

**Comment by oleq** [comment]
I'm also for removing `(add|remove|get)Child` methods from `View` since they're gone from `Controller`.

**Comment by Reinmar** [comment]
> I think the following is very confusing
>
> `this.children.get( 'buttons' ).get( 0 );`
>
> so I'm for this.collections rather than this.children. But if you insist, it's easy to change.

Hm... OK, let's keep what we've got now. If someone wants to shorten the path it's also possible to create ctrl's property and have:

```
this.buttons.get( 0 )
```

**Comment by Reinmar** [comment]
Made some minor refactoring and reported #118. And I think we're ready. Let's see this in action :D.

**Pull request merged**